### PR TITLE
AP_Motors: add arming check for VOLT_MIN and VOLT_MAX

### PR DESCRIPTION
**Pull request by IamPete1**
This adds a arming check to the thrust linearization class for bat min and bat max. They must be configured if there is a none zero voltage reading from the selected monitor. This relies on battery monitor returning 0 for invalids battery index's. The arming check can be bypassed by setting the battery index to -1. Unfortunately there is not enough room to fit that in the pre arm message.

This pre-arm is shown on both copter and plane. @bnsgeyer we might want to look at enabling it on heli in the future too.

### CODE REVIEW

**Comment by IamPete1** [comment]
Your right this probably will cause a pre-arm for lots of users, but unless there setup you can't really tune the vehicle properly. You don't have to setup the functionality in order to get past the pre-arm you can set the battery index to -1.

Note that this will not give the pre arm on vehicles that do not have a battery monitor setup.

We would need to add this to the wiki so users can work out how to resolve the issue.

**Comment by Hwurzburg** [comment]
still not a fan of this.....I think it will cause ill will from many how are happily flying now without having this setup and upgrade only to get a prearm, whether or not having these params set is the optimum configuration or not

**Review by lthall — COMMENTED** [review]
This looks fine to me. I am not sure why this will cause an issue for users though. I thought the defaults would mean battery_voltage_config_valid would return true because AP_MOTORS_BAT_VOLT_MAX_DEFAULT = 0.0f.

What am I missing?

**Comment by Hwurzburg** [comment]
yes, I thought that they had defaults at zero, hence the usefulness of this prearm....but the defaults for MAX/MIN would not cause a pre-arm...so that now begs the question: if defaults are set, and someone changing them would usually NOT put them at in this very unusual config that this checks...and if they did, it would not cause a crash, just poor tuning...like them changing many other params incorrectly.....why add the code for the prearm at all?

**Comment by lthall** [comment]
> if defaults are set, and someone changing them would usually NOT put them at in this very unusual config that this checks...and if they did, it would not cause a crash, just poor tuning...like them changing many other params incorrectly.....why add the code for the prearm at all?

I don't know if it is worth a prearm or what the cost of adding a prearm is. However, if someone has set this such that the prearm is activated then it would be a very important thing for them to know about because this would have a significant negative impact on their tune. Without voltage compensation the PID's have to be set to between 60% and 70% of the optimal values.

I don't know if this is worth a prearm check but I would say anybody that this hits will benefit from it more than they are inconvenienced.

**Comment by IamPete1** [comment]
> This looks fine to me. I am not sure why this will cause an issue for users though. I thought the defaults would mean battery_voltage_config_valid would return true because AP_MOTORS_BAT_VOLT_MAX_DEFAULT = 0.0f.

The default of 0 will trigger the pre-arm. 0 is not greater than 0. I could change it to `is_positive` which might be a little clearer.

**Comment by Hwurzburg** [comment]
bugger....I was looking at sitl defaults which would not generate the prearm...normal code loads will have defaults that trigger the prearm....so my objection to causing prearms on existing flying configurations when upgrading stands...
however, having a non blocking pream warning would be okay...cant you issue a non blocking message in the prearm check loop?

**Comment by lthall** [comment]
I would not do the arming check if they have not set up the motor min max but only if they set it up incorrectly, ie max < min.

**Comment by IamPete1** [comment]
I have marked the for dev call, I think everyone agrees that vehicles should have this setup, but I suspect the answer will be that a arming check is a step too far

File: libraries/AP_Motors/AP_Motors_Thrust_Linearization.cpp [hunk 1]
```diff
@@ -197,3 +203,25 @@ float Thrust_Linearization::get_compensation_gain() const
...
+bool Thrust_Linearization::arming_checks(size_t buflen, char *buffer) const
+{
+    if (get_spin_min() > 0.3) {
+        hal.util->snprintf(buffer, buflen, "%sSPIN_MIN too high %.2f > 0.3", AP_MOTORS_PARAM_PREFIX, get_spin_min());
+        return false;
+    }
+
+    if (is_positive(AP::battery().voltage(batt_idx)) && !battery_voltage_config_valid()) {
+        // If there is a valid voltage reading battery voltage scaling should be setup
+        hal.util->snprintf(buffer, buflen, "Set %sBAT_VOLT_MIN and %sBAT_VOLT_MAX", AP_MOTORS_PARAM_PREFIX, AP_MOTORS_PARAM_PREFIX);
```
Comment (peterbarker) [inline libraries/AP_Motors/AP_Motors_Thrust_Linearization.cpp:222]: I wonder if the compiler is smart enough to 
```suggestion
        hal.util->snprintf(buffer, buflen, "Set " AP_MOTORS_PARAM_PREFIX " BAT_VOLT_MIN and " AP_MOTORS_PARAM_PREFIX "BAT_VOLT_MAX");
```

**Review by rmackay9 — CHANGES_REQUESTED** [review]
As discussed on the dev call, I think this raises the bar too high for new users who just want to get their vehicle flying.  Instead we should work on GCS or web tool to to make it easier for users to set these values properly.

**Comment by IamPete1** [comment]
The conclusion is this arming check is a step to far.

**Pull request closed without merging**